File: python/dynare_py/simulation.py
"""
Impulse response functions and stochastic simulation.
"""

from __future__ import annotations

from typing import Dict, List, Optional, Sequence, Union

import numpy as np
import pandas as pd

from .gensys import GensysResult


# ------------------------------------------------------------------ #
# IRF                                                                  #
# ------------------------------------------------------------------ #

def compute_irf(
    result: GensysResult,
    shock_index: int = 0,
    shock_size: float = 1.0,
    periods: int = 40,
    var_names: Optional[List[str]] = None,
) -> pd.DataFrame:
    """
    Compute impulse responses to a one-unit shock.

    Parameters
    ----------
    result : GensysResult from gensys()
    shock_index : which column of R (the shock) to hit
    shock_size : scale (default 1.0, i.e. 1 std dev)
    periods : number of periods (including impact period 0)
    var_names : labels for columns; defaults to y0, y1, ...

    Returns
    -------
    DataFrame, shape (periods, n_variables)
    """
    if not result.success:
        raise ValueError("Cannot compute IRF: model has no unique solution.")

    n = result.T.shape[0]
    n_e = result.R.shape[1]

    if var_names is None:
        var_names = [f"y{i}" for i in range(n)]

    irf = np.zeros((periods, n))
    # Impact period: shock hits
    eps = np.zeros(n_e)
    eps[shock_index] = shock_size
    irf[0] = result.R @ eps + result.const

    # Subsequent periods: propagate
    for t in range(1, periods):
        irf[t] = result.T @ irf[t - 1] + result.const

    return pd.DataFrame(irf, columns=var_names, index=pd.RangeIndex(periods, name="period"))


def compute_all_irfs(
    result: GensysResult,
    shock_names: Optional[List[str]] = None,
    var_names: Optional[List[str]] = None,
    periods: int = 40,
) -> Dict[str, pd.DataFrame]:
    """
    Compute IRFs for all shocks.  Returns {shock_name: DataFrame}.
    """
    n_e = result.R.shape[1]
    if shock_names is None:
        shock_names = [f"eps{i}" for i in range(n_e)]

    return {
        shock_names[i]: compute_irf(
            result, shock_index=i, periods=periods, var_names=var_names
        )
        for i in range(n_e)
    }


# ------------------------------------------------------------------ #
# Stochastic simulation                                                #
# ------------------------------------------------------------------ #

def simulate(
    result: GensysResult,
    shock_cov: Optional[np.ndarray] = None,
    periods: int = 200,
    burn: int = 50,
    seed: Optional[int] = None,
    var_names: Optional[List[str]] = None,
) -> pd.DataFrame:
    """
    Simulate the model forward by drawing shocks.

    Parameters
    ----------
    result : GensysResult
    shock_cov : (n_e, n_e) covariance matrix of eps_t.
                If None uses identity (all shocks iid N(0,1)).
    periods : number of periods to return (after burn-in)
    burn : burn-in periods discarded
    seed : random seed

    Returns
    -------
    DataFrame, shape (periods, n_variables)
    """
    if not result.success:
        raise ValueError("Cannot simulate: model has no unique solution.")

    rng = np.random.default_rng(seed)
    n = result.T.shape[0]
    n_e = result.R.shape[1]

    if shock_cov is None:
        shock_cov = np.eye(n_e)

    L = np.linalg.cholesky(shock_cov)   # lower-triangular factor

    total = burn + periods
    out = np.zeros((total, n))
    y = np.zeros(n)

    for t in range(total):
        eps = L @ rng.standard_normal(n_e)
        y = result.T @ y + result.R @ eps + result.const
        out[t] = y

    if var_names is None:
        var_names = [f"y{i}" for i in range(n)]

    return pd.DataFrame(
        out[burn:], columns=var_names, index=pd.RangeIndex(periods, name="period")
    )


# ------------------------------------------------------------------ #
# Forecast error variance decomposition (FEVD)                         #
# ------------------------------------------------------------------ #

def fevd(
    result: GensysResult,
    shock_cov: Optional[np.ndarray] = None,
    horizon: int = 20,
    var_names: Optional[List[str]] = None,
    shock_names: Optional[List[str]] = None,
) -> Dict[str, pd.DataFrame]:
    """
    Forecast error variance decomposition.

    Returns
    -------
    dict with 'variance' and 'share' DataFrames, both shape (horizon, n_vars).
    Each column = total variance / share for that variable at each horizon.
    Plus per-shock DataFrames under key 'by_shock'.
    """
    if not result.success:
        raise ValueError("Cannot compute FEVD: model has no unique solution.")

    n = result.T.shape[0]
    n_e = result.R.shape[1]

    if shock_cov is None:
        shock_cov = np.eye(n_e)
    if var_names is None:
        var_names = [f"y{i}" for i in range(n)]
    if shock_names is None:
        shock_names = [f"eps{i}" for i in range(n_e)]

    T, R = result.T, result.R
    L = np.linalg.cholesky(shock_cov)
    RL = R @ L   # Cholesky-scaled impact

    # var_by_shock[j][h] = contribution of shock j to variance of each variable at h
    var_total = np.zeros((horizon, n))
    var_by_shock = {s: np.zeros((horizon, n)) for s in shock_names}

    T_power = np.eye(n)  # T^0 at h=0
    for h in range(horizon):
        for j, sname in enumerate(shock_names):
            r = T_power @ RL[:, j]   # response at h to unit Chol shock j
            contrib = r ** 2
            var_by_shock[sname][h] += contrib
            var_total[h] += contrib
        T_power = T @ T_power

    # Shares
    var_share_by_shock = {}
    for sname in shock_names:
        share = np.where(var_total > 0, var_by_shock[sname] / var_total, 0.0)
        var_share_by_shock[sname] = pd.DataFrame(share, columns=var_names)

    return {
        "variance": pd.DataFrame(var_total, columns=var_names),
        "by_shock": {s: pd.DataFrame(var_by_shock[s], columns=var_names) for s in shock_names},
        "share": var_share_by_shock,
    }

File: julia/DynareJL/src/simulation.jl
"""
IRF and stochastic simulation for DynareJL.
"""

using Statistics
using LinearAlgebra

# ------------------------------------------------------------------ #
# IRF                                                                  #
# ------------------------------------------------------------------ #

"""
    compute_irf(result, shock_index; shock_size=1.0, periods=40, var_names=nothing)
              -> Matrix{Float64}  (periods × n_vars)

Compute the impulse response to a single shock.
Returns a matrix; rows = periods (0-indexed), columns = variables.
"""
function compute_irf(
    result::GensysResult,
    shock_index::Int = 1;
    shock_size::Float64 = 1.0,
    periods::Int = 40,
    var_names::Union{Vector{String},Nothing} = nothing,
)
    result.success || error("Cannot compute IRF: model has no unique solution.")

    n   = size(result.T, 1)
    n_e = size(result.R, 2)

    irf = zeros(periods, n)
    eps = zeros(n_e); eps[shock_index] = shock_size

    irf[1, :] = result.R * eps .+ result.const_vec
    for t in 2:periods
        irf[t, :] = result.T * irf[t-1, :] .+ result.const_vec
    end
    return irf
end

"""
    compute_all_irfs(result; shock_names, var_names, periods=40)
                  -> Dict{String, Matrix{Float64}}

Compute IRFs for all shocks.
"""
function compute_all_irfs(
    result::GensysResult;
    shock_names::Union{Vector{String},Nothing} = nothing,
    var_names::Union{Vector{String},Nothing} = nothing,
    periods::Int = 40,
)
    n_e = size(result.R, 2)
    snames = shock_names === nothing ? ["eps$i" for i in 1:n_e] : shock_names
    return Dict(
        snames[i] => compute_irf(result, i; periods=periods, var_names=var_names)
        for i in 1:n_e
    )
end

# ------------------------------------------------------------------ #
# Stochastic simulation                                                #
# ------------------------------------------------------------------ #

"""
    simulate(result; shock_cov=I, periods=200, burn=50, seed=nothing, var_names=nothing)
           -> Matrix{Float64}  (periods × n_vars)

Simulate the model by drawing shocks.

# Arguments
- `shock_cov` : (n_e × n_e) covariance; defaults to identity
- `periods`   : simulation length after burn-in
- `burn`      : discarded burn-in periods
- `seed`      : random seed
"""
function simulate(
    result::GensysResult;
    shock_cov::Union{AbstractMatrix,Nothing} = nothing,
    periods::Int = 200,
    burn::Int = 50,
    seed::Union{Int,Nothing} = nothing,
    var_names::Union{Vector{String},Nothing} = nothing,
)
    result.success || error("Cannot simulate: model has no unique solution.")

    n   = size(result.T, 1)
    n_e = size(result.R, 2)

    Σ = shock_cov === nothing ? Matrix{Float64}(I, n_e, n_e) : Matrix{Float64}(shock_cov)
    L = cholesky(Σ).L

    if seed !== nothing
        Random.seed!(seed)
    end

    total = burn + periods
    out = zeros(total, n)
    y   = zeros(n)

    for t in 1:total
        eps = L * randn(n_e)
        y   = result.T * y .+ result.R * eps .+ result.const_vec
        out[t, :] = y
    end

    return out[burn+1:end, :]
end

# ------------------------------------------------------------------ #
# Forecast error variance decomposition                                #
# ------------------------------------------------------------------ #

"""
    fevd(result; shock_cov=I, horizon=20, shock_names=nothing, var_names=nothing)
       -> NamedTuple(variance, share_by_shock)

Returns:
- `variance`      : (horizon × n_vars) total forecast error variance
- `share_by_shock`: Dict mapping shock name -> (horizon × n_vars) share matrix
"""
function fevd(
    result::GensysResult;
    shock_cov::Union{AbstractMatrix,Nothing} = nothing,
    horizon::Int = 20,
    shock_names::Union{Vector{String},Nothing} = nothing,
    var_names::Union{Vector{String},Nothing} = nothing,
)
    result.success || error("Cannot compute FEVD: model has no unique solution.")

    n   = size(result.T, 1)
    n_e = size(result.R, 2)

    Σ = shock_cov === nothing ? Matrix{Float64}(I, n_e, n_e) : Matrix{Float64}(shock_cov)
    L = cholesky(Σ).L
    RL = result.R * L   # (n × n_e)

    snames = shock_names === nothing ? ["eps$i" for i in 1:n_e] : shock_names

    var_total = zeros(horizon, n)
    var_shock = Dict(s => zeros(horizon, n) for s in snames)

    T_power = Matrix{Float64}(I, n, n)  # T^0
    for h in 1:horizon
        for j in 1:n_e
            r     = T_power * RL[:, j]
            contrib = r .^ 2
            var_shock[snames[j]][h, :] .+= contrib
            var_total[h, :]            .+= contrib
        end
        T_power = result.T * T_power
    end

    share_by_shock = Dict(
        s => [var_total[h,v] > 0 ? var_shock[s][h,v] / var_total[h,v] : 0.0
              for h in 1:horizon, v in 1:n]
        for s in snames
    )

    return (variance=var_total, share_by_shock=share_by_shock)
end
